examples/chat2-reliable/chat.go

```diff
@@ -137,20 +151,65 @@ func (c *Chat) receiveMessages() {
 		case <-c.ctx.Done():
 			return
 		case value := <-c.C:
-
 			msgContentTopic := value.Message().ContentTopic
 			if msgContentTopic != c.options.ContentTopic {
 				continue // Discard messages from other topics
 			}
 
 			msg, err := decodeMessage(c.options.ContentTopic, value.Message())
 			if err == nil {
-				// send valid messages to the UI
-				c.ui.ChatMessage(int64(msg.Timestamp), msg.Nick, string(msg.Payload))
+				c.processReceivedMessage(msg)
 			}
 		}
 	}
 }
+
+func (c *Chat) processReceivedMessage(msg *pb.Message) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Update Lamport timestamp
+	if msg.LamportTimestamp > c.lamportTimestamp {
+		c.lamportTimestamp = msg.LamportTimestamp
+	}
+	c.lamportTimestamp++
+
+	// Check causal dependencies
+	if c.checkCausalDependencies(msg) {
+		// Process the message
+		c.ui.ChatMessage(int64(c.lamportTimestamp), msg.SenderId, msg.Content)
+
+		// Update bloom filter
+		c.updateBloomFilter(msg.MessageId)
+
+		// Add to message history
+		c.addToMessageHistory(msg)
+	} else {
+		// Add to incoming buffer
+		c.incomingBuffer = append(c.incomingBuffer, msg)
+	}
+}
+
+func (c *Chat) checkCausalDependencies(msg *pb.Message) bool {
+	for _, depID := range msg.CausalHistory {
+		if !c.bloomFilter.Test([]byte(depID)) {
+			return false
+		}
```

Comment:
Causal dependencies should preferably be checked against the actual local history (i.e. `messageHistory`), not the bloom filter over the history. For one, the bloom filter should only be over _recent_ history to keep the number of included items low (thus reducing fp rate) whereas causal history should cover "everything". Causal history should also be a deterministic mechanism, while bloom filter is probabilistic.